connlistener/connlistener.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// package connlistener provides a net.Listener to which one can hand
+// connections directly.
+package connlistener
+
+import (
+	"net"
+	"sync/atomic"
+	"syscall"
+)
+
+// Listener is a net.Listener to which one can manually hand new connections
+// using the HandleConn() method.
+type Listener interface {
+	net.Listener
+
+	// HandleConn makes the specified conn available to callers of this
+	// Listener's Accept() function. It takes the remoteAddr of the conn in
+	// case the supplied net.Conn does not provide the right value for
+	// RemoteAddr(). This method returns an error if the Listener has been
+	// closed.
+	HandleConn(conn net.Conn, remoteAddr net.Addr) error
+}
+
+type connListener struct {
+	ch     chan net.Conn
+	closed atomic.Bool
+}
+
+// New creates a Listener.
+func New() Listener {
+	return &connListener{
+		ch: make(chan net.Conn, 10), // we use a small buffer to avoid blocking callers to ConnAvailable
```

Comment:
The aggregate of your comments about ConnListener made me realize that it's not an abstraction that's ready for public consumption. Also, it's only used in tailfs. Consequently, I've moved it into the tailfs package.

I made the channel unbuffered so that it back-pressures, as that seems like the right behavior here.

Reply:
I made this unbuffered.